pkg/pillar/cmd/baseosmgr/handlebaseos.go

```diff
@@ -1018,3 +1021,107 @@ func handleOtherPartRebootReason(ctxPtr *baseOsMgrContext, status *types.BaseOsS
 	}
 }
+
+//isImageInErrorState returns true if we try to update to IMGB without success
+//also returns ZbootStatus
+func isImageInErrorState(ctxPtr *baseOsMgrContext) (bool, *types.ZbootStatus) {
+	otherPartName := zboot.GetOtherPartition()
+	partStatus := getZbootStatus(ctxPtr, otherPartName)
+	if partStatus == nil {
+		log.Functionf("No other partition status for %s",
+			otherPartName)
+		return false, nil
+	}
+	shortVerOtherPart := partStatus.ShortVersion
+	if shortVerOtherPart == "" {
+		log.Functionf("Other partition has no version")
+		return false, partStatus
+	}
+	if partStatus.PartitionState != "inprogress" {
+		log.Functionf("Other partition state %s not inprogress",
+			partStatus.PartitionState)
+		return false, partStatus
+	}
+	baseOSConfig := lookupBaseOsConfigByVersion(ctxPtr, shortVerOtherPart)
+	if baseOSConfig == nil {
+		log.Functionf("Cannot found BaseOsConfig with %s version; ignoring RetryUpdateCounter",
+			shortVerOtherPart)
+		return false, partStatus
+	}
+	if !baseOSConfig.Activate {
+		log.Functionf("BaseOsConfig %s has no activate; ignoring RetryUpdateCounter", baseOSConfig.Key())
+		return false, partStatus
+	}
+	return true, partStatus
+}
+
+// handleUpdateRetryCounter checks
+// if current partition is not active: just update configUpdateRetry
+// if other partition is in failed state: if retryUpdateCounter changed: save current counters and initiate re-update
+// in other case update configUpdateRetry and currentUpdateRetry, save them and publish
+func handleUpdateRetryCounter(ctxPtr *baseOsMgrContext, retryUpdateCounter uint32) {
+	curPartName := zboot.GetCurrentPartition()
+	partStatus := getZbootStatus(ctxPtr, curPartName)
+	if partStatus == nil {
+		log.Warnf("handleUpdateRetryCounter: No current partition status for %s; failed RetryUpdateCounter",
+			curPartName)
+		return
+	}
+	if partStatus.PartitionState != "active" {
+		if ctxPtr.configUpdateRetry == retryUpdateCounter {
+			log.Functionf("No change in retryUpdateCounter: %d", retryUpdateCounter)
+			return
+		}
+		log.Noticef("handleUpdateRetryCounter: configUpdateRetry change: %d to %d",
+			ctxPtr.configUpdateRetry, retryUpdateCounter)
+		ctxPtr.configUpdateRetry = retryUpdateCounter
+		return
+	}
+	if failed, failedPartStatus := isImageInErrorState(ctxPtr); failed {
+		if ctxPtr.configUpdateRetry == retryUpdateCounter {
+			log.Functionf("No change in retryUpdateCounter: %d", retryUpdateCounter)
+			return
+		}
+		log.Noticef("handleUpdateRetryCounter: configUpdateRetry change: %d to %d",
+			ctxPtr.configUpdateRetry, retryUpdateCounter)
+		ctxPtr.configUpdateRetry = retryUpdateCounter
+		saveRetryUpdateCounter(false, ctxPtr.currentUpdateRetry)
+		// save it to avoid loop of re-upgrade - failing - re-upgrade
+		saveRetryUpdateCounter(true, ctxPtr.configUpdateRetry)
+		log.Noticef("UpdateRetry from %s to %s",
+			partStatus.ShortVersion, failedPartStatus.ShortVersion)
+		zboot.SetOtherPartitionStateUpdating(log)
```

Comment:
I was assuming you'd trigger a full redownload, but as long as the checks in isImageInErrorState are tight this is ok.

Reply:
But where is the actual reboot into the other image?

Reply:
If see the above code matches the forcefallback code, so does that mean that the reboot happens automagically somewhere?

Reply:
Right, they are the same 

> If see the above code matches the forcefallback code, so does that mean that the reboot happens automagically somewhere?

Right, they are the same on that step, so, actions are driving by nodeagent.

>I was assuming you'd trigger a full redownload, but as long as the checks in isImageInErrorState are tight this is ok.

Downloading is already done on this stage, when we add something into BaseOsConfig. Here I assume that we are failed on testing stage for some reason.